system/View/Table.php

```diff
@@ -207,7 +213,38 @@ public function setEmpty($value)
      */
     public function addRow()
     {
-        $this->rows[] = $this->_prepArgs(func_get_args());
+        $tmpRow = $this->_prepArgs(func_get_args());
+
+        if ($this->syncRowsWithHeading && ! empty($this->heading)) {
+            // each key has an index
+            $keyIndex = array_flip(array_keys($this->heading));
+
+            // figure out which keys need to be added
+            $missingKeys = array_diff_key($keyIndex, $tmpRow);
+
+            // Remove all keys which don't exist in $keyIndex
+            $tmpRow = array_filter($tmpRow, static fn ($k) => array_key_exists($k, $keyIndex), ARRAY_FILTER_USE_KEY);
+
+            // add missing keys to row, but use $this->emptyCells
+            $tmpRow = array_merge($tmpRow, array_map(fn ($v) => ['data' => $this->emptyCells], $missingKeys));
+
+            // order keys by $keyIndex values
+            uksort($tmpRow, static fn ($k1, $k2) => $keyIndex[$k1] <=> $keyIndex[$k2]);
+        }
+        $this->rows[] = $tmpRow;
+
+        return $this;
+    }
+
+    /**
+     * Set to true if each row column should be synced by keys defined in heading.
+     *
+     * If a row has a key which does not exist in heading, it will be filtered out
+     * If a row does not have a key which exists in heading, the field will stay empty
+     */
+    public function setSyncRowsWithHeading(bool $orderByKey): Table
```

Comment:
This is not bad, but other methods do not have the return type. And we would like to add `static` when we drop PHP 7.4 support.

```suggestion
    public function setSyncRowsWithHeading(bool $orderByKey)
```